Proposal to manage kyma components using  kyma cli (#487)
kyma-project/community: collaboration/extending-kyma-cli.md

> # Extend Kyma CLI

Author: updated.

> # Extend Kyma CLI

Reviewer: Reasoning is much better now, please reflect it in the title in PR name as well, what is the actual goal? It is not extending the CLI.

> Kyma cli has been used to provision a cluster and install kyma into. We should look into extend the current functionality set such that we can provide functionality to access various features of kyma.

Author: updated with new text.

> The extension of functionality would enable people with no or less kubernetes knowledge can use kyma without any issue.

Author: updated

## Security

> Kyma cli should support RBAC. This would be realised by downloading the requisite kubeconfig file. The admins should be able to use more infrastructure commands and developers should be able to use the kyma cluster with restricted rights.

Author: removed.

## Architecture

  We should follow the modular approach with each component having its own library. This library can be used inside the kyma cli for acessing various components. The library contains all the implementation. The kyma cli should invoke this library. We can take inspiration from hydroform and implement it in similar way. Currently we have following modules in kyma:

>   1. Lambdas

Reviewer: Will the project templating fall into the function module as well?

Author: yes. Since its a feature for functions.

  2. Service Catalog
  3. Events
  4. Application Connector
  5. APIs

## Command structure

  The command structure used by cloud providers is following:

  ```bash
  <cliName> <functionality> <action>
  #example aws s3 ls
  ```

  The current kyma-cli also uses the same pattern eg.

  ```bash

>   kyma provision

Author: updated.

  or
  kyma install
  ```

### Lambdas

For the serverless component we should start with supporting following commands

> #### Local workspace setup

Author: updated.

> 1. Creation of project locally

Author: updated.

>   `kyma lambda init <lambda-name> --runtime <runtime-name> -n <namespace>` or `kyma lambda init <lambda-name> --runtime <runtime-name> -n <namespace> -p <path>`

Author: updated.

  This should create directoy with following content based on the runtime:

      ```bash
        <lambda-name>
        ├── .vscode
        │   └── launch.json
        ├── deployment
        │   └── deployment.yaml
        ├── local
        │   └── index.js
        └── code
          ├── config.yaml
          ├── handler.js
          └── package.json
      ```

>     > When path `-p` is passed then it should create in desitnation path else it use the workspace set in the yaml file.

Author: fixed.

>     * launch.json consists of settings for the debugging

Author: I have added it for now but we can discuss if we should be IDE agnostic.

    * deployment.yaml consists of the yaml that has been deployed on the kyma cluster.
    * `index.js` consists of the server code with handler pointing to `handler.js`

>     * config.yaml can be used as info file with current configurations like:

Author: updated.

>       - name

Reviewer: So the config.yaml mainly contains all properties of the function resource, so that the user do not need to understand the k8s specific way of expressing things, right? 
Can you add a discussion about pro and cons and what it would mean to use a function resource file instead which can be deployed also without kymaCLI. I think that discussion we will have a lot later so we should have an arguing.

Author: updated.

      - namespace
      - events subscribed
      - environment variables
      - Api name to expose with
      - replicas
      - min/max memory
      - min/max cpu
      - runtime
      - debug command (to be used with telepresence. This can help if we want to have different runtimes)

>     * handler.js where the developer can write his logic to test

Author: updated.

    * package.json file with dependencies.

    >The templates generated should be placed a directory in the kyma cli. In the future if we are supporting more runtimes we can move it to separate repo.

2. Testing/debugging it locally

We can introduce commands to test the lambda locally and also if the developer wants to debug the code. This is useful in case when we have some event payloads that are not easy to recreate.

> `kyma lambda debug <lambda-name>`

Reviewer: Please explain a bit more what this command should do. Local debugging means to start via kymaCLI the lambda locally and attach a debugger to it?

Author: updated.

> We can use telepresence for local debugging. We can either use --swap-deployment or --new-deployment option

> #### Working with deployed kyma cluster

Reviewer: Will be good to have a separate listing of the commands supporting the config.yaml to understand what plays in to the main lambda project scenario. So for event binding I would also pass the project folder as argument?

It is nice to see the overall picture on what might be supported long-term but we should focus more on the lambda project scenario. What commands a user will use e2e when having such lambda project? WHat are the typical flows and with that the command chains?

Author: updated.

We should allow deployment of the lambdas on a k8s cluster. For this we should use the service account role. Following commands should be supported

##### CRUD

`kyma lambda create <lambda-name> -n <namespace> --runtime <runtime> --min-mem <mb> --max-mem <mb> --min-cpu <cpu> --max-cpu <cpu> --min-replicas 1 --max-replicas 1 --src-code <path-to-src-code> --dependecies <path-to-dep-file>`

`kyma lambda create -p <path-to-lambda-dir>`

> > This would create a lambda after reading the deployment.yaml from the path provided and also would read `config.yaml` to read the config like events, api etc.

Reviewer: Why deployment.yaml, shouldn't be the config.yaml enough?

Author: fixed.

`kyma lambda update <lambda-name> -n <namespace> --runtime <runtime> --min-mem <mb> --max-mem <mb> --min-cpu <cpu> --max-cpu <cpu> --min-replicas 1 --max-replicas 1 --src-code <path-to-src-code> --dependecies <path-to-dep-file>`

`kyma lambda update -p <path-to-lambda-dir>`

>  > Update an existing lambda with the deployment.yaml passed.

Reviewer: Why deplyoment.yaml?

Author: fixed,

# Extend Kyma CLI

Kyma cli has been used for cluster management. We should look into how we can introduce support for management of components on the kyma cluster. This would bring in following benefits:

> * It would enable users who have limited or no kuberenetes knowledge to use kyma.

Author: fixed.

* Provide an abstraction on complex kubectl command chains.
* Could be easily used into CI/CD pipelines where they can have automated way of configuring kyma cluster to have functions deployed via CI/CD pipeline

## Architecture

> We should follow the modular approach with each component having its own library. This library can be used inside the kyma cli for acessing various components. The library contains all the implementation. The kyma cli should invoke this library. We can take inspiration from hydroform and implement it in similar way. Currently we have following modules in kyma:

Reviewer 2: Nice 👍 Having a sdk would be great and could replace a lot of our logic in tests.

Reviewer 3: The sdk can evolve in time, covering aspects like: management of Kyma components, rapid application development (like Dapr's languages sdk) and of course local/remote development.

  1. Functions
  2. Service Catalog
  3. Events
  4. Application Connector
  5. APIs

## Command structure

The command structure used by cloud providers is following:

```bash
<cliName> <functionality> <action>
#example aws s3 ls

# Functions
kyma function <action>

# Events
kyma events <action>

# Application Connector

> kyma application <events>

Author: fixed.

# Apis
kyma apis <events>
```

> ### functions

Author: fixed.

For the serverless component we should start with supporting following commands

#### Local workspace setup

> For the local workspace setup developer should be provied with a project containing basic working function example. This should be deployable on to an existing kyma cluster using the kyma cli. It should also be possible to test this code locally by means of make files. User should be provded with some readme where he can understand the project structure and also the configuration examples. It should also contain commands to deploy and debug the function on to kyma cluster and also how to run locally.

Author: fixed.

##### Creation of project locally

This could be used for the user who is new to kyma and kubernetes in general. He should be able to have an example function created easily. We can have following commands.

```bash
  kyma function init <function-name> --language <language-name>

  kyma function init <function-name> --language <language-name> -n (optional) <namespace> -p <path>

  kyma function init <function-name> --language <language-name> --from-git <path-to-function-code>
```

> > The namespace should be optional. When not passed it can be set to `default`. When the user has path to the function code (residing in git) we should generate the scaffolding around it (if its not existing already).

Reviewer 4: I think from the first command, it is obvious that these flags are optional. No need to repeat that.

Author: fixed.

> When path `-p` is passed then it should create in desitnation path else it use the current directory.

Reviewer 4: ```suggestion
When path `-p` is passed then it should create the project in the designated path, otherwise, it uses the current directory.
```

Author: fixed.

> The language corresponds the language in which the function code would be written.

Author: restructured.

This should create directoy structure with following content based on the language passed (assuming vscode as the IDE).:

```bash
  <function-name>
  ├── .vscode
  │   └── launch.json
  ├── deployment
  │   └── function.yaml
  ├── local
  │   └── <server-code.extension>
  └── Readme.md
  └── src
    ├── config.yaml
    ├── <function-code.extension>

>     └── <dependecies.extension>

Author: fixed.

```

* launch.json consists of settings for the local debugger on vscode

> * function.yaml consists of the yaml of the function cr that has been deployed on the kyma cluster. This would be generated once function has been deployed.

Author: restructed.

> * `<server-code.extension>` consists of the server code with handler pointing to <function-code.extension>. The `extension` of file and `server-code` is with repsect to the language passed in init command

Author: yup

* config.yaml can be used as info file with current configurations like:
  * name
  * namespace
  * events subscribed
  * environment variables
  * Api name to expose with
  * replicas
  * min/max memory
  * min/max cpu
  * runtime
  * debug command (to be used with telepresence. This can help if we want to have different runtimes)
* <function-code.extension> where the developer can write his logic to test. The `extension` is language specific.

> * `dependecies.extension` file with list of dependencies. eg. `package.json` for node and `go.mod and go.sum` for golang dependencies

Author: yeah. We should not be restricted to node.

> >The templates generated should be placed a directory in the kyma cli. We should have templates for each language that can be supported. In the future if we are supporting more runtime we can move it to separate repo.

Reviewer 4: I thought they would reside on a new hydroform package.

Author: Nice catch thanks 👍

> >The templates generated should be placed a directory in the library for the functions. We should have templates for each language that can be supported. In the future if we are supporting more runtime we can move it to separate repo.

Author: It means its part of the sdk.

##### config.yaml

This file basically enables developers to configure their functions. With init command this file is created and it basically has some default values. The readme files provide information about various options present in this file

Developer can modify this file to deploy or update a function on to the kyma cluster. This would enable the developer not only tune the function (by modifying replica, memory cpu), but also provide an overview of the current triggers for the function.

Usually one does not change the config so often during development compared to code. So our aim should be to make deployment and testing the function simpler. So having a config file taking care of configuration would make deployment much simpler.

Having such a config file would help developers with limited or no kubernetes knowledge to get themselves aquainted with various configuration options for function and concentrate on the business logic for writing the functions.

> Additionally it can be used automated deployments of functions. As this file can be read by kyma cli to deloy function with required configurations.

Author: fixed.

Example structure of config.yaml

```yaml
name: <function_name>
namespace: <namespace> #default by default
events:
  - event1
  - event2
environment-vars:
  - foo:bar
api:
  url: <url>
  actions:
    - GET
    - POST
  security:
    - JWT
replicas: 1
memory:
  min:
  max:
cpu:
  min:
  max:
runtime:
debug-string:
```

##### Template files

###### <function-code.extension>

  Out of the box from the init command we should give an easy example which can be extended by the developer later. The example should be possible to be deployed directly on the kyma cluster. Here below there is a node example

  ```js
    module.exports = {
      main: function (event, context) {
        return 'Hello World';
      }
    };
  ```

###### <dependecies.extension>

We should generate an empty package.json file which can be edited by developer as and when needed.

###### index.js

We should provide `index.js`(which would start the server and serve the handler function) out of the box which can be used by the developer to test his code locally.

###### readme

> We should also provide a readme.md which would consist of following:

Reviewer 4: You can mention "instructions" here and delete it from the following list items.

Author: fixed.

* Instructions how he can deploy the function into the cluster.
* Instructions how he can run the function locally.
* Instructions on how to use the `config.yaml` file
* Instructions on various folder structure.
* Instructions on how to test debug it locally like using telepresence.

> ##### Testing/debugging it locally

Reviewer: You actually mean debugging a remote running function on local machine. When reading it first I always mix it up with simple local debugging, so starting it local and debugging it local. It should be stated more clear that this is covered differently.

*Local*

We can introduce commands to test the function locally and also if the developer wants to debug the code. These can be achieved through make file.

*Using Events from remote cluster*

> Sometimes the event payloads can be complex to be created locally. For such cases its nice to have the events on remote cluster triggering the code on the workstation. This would enable the developer to debug his code on the workstation using events being sent to the remote cluster. We can introduce command for example below:

Author: how about `kyma event trigger --topic {"foo":"bar"} --data {"foo": "bar"}`

> `kyma function debug <function-name> -n <namespace>`

Reviewer 3: Allowing a distributed development environment is very important for application development (in our case writing a "function"). The sdk should support this, based on the eventing implementation which should be able to make a connection between the remote cluster and the local "eventing" part (the event generated on remote cluster should be propagated to the local "eventing"). We need to have something similar with kubectl port-forward but adapted to our context ( eventing-forward ...)

Author: we should discuss this in eventing section how can we achieve this local eventing communicating with remote cluster.

#### Working with deployed kyma cluster

We should allow deployment of the functions on a k8s cluster. Following commands should be supported

##### CRUD

##### Commands that use config

`kyma function create -p <path-to-function-dir>`
> This would create a function after reading the `config.yaml` from the path provided and to read the config like events, api etc.

`kyma function update -p <path-to-function-dir>`
> Update an existing function with the `config.yaml` present in the directory.

##### Expanded command with switches

`kyma function create <function-name> -n <namespace> --runtime <runtime> --min-mem <mb> --max-mem <mb> --min-cpu <cpu> --max-cpu <cpu> --min-replicas 1 --max-replicas 1 --src-code <path-to-src-code> --dependecies <path-to-dep-file>`

`kyma function update <function-name> -n <namespace> --runtime <runtime> --min-mem <mb> --max-mem <mb> --min-cpu <cpu> --max-cpu <cpu> --min-replicas 1 --max-replicas 1 --src-code <path-to-src-code> --dependecies <path-to-dep-file>`

`kyma function delete <function-name> -n <namespace>`

`kyma function get <function-name> -n <namespace>`

*Labels*

`kyma function label <function-name> ["foo=bar",..] -n <namespace>`

*Expose*

> `kyma function trigger <function-name> --http --secure(optional) -n <namespace>`

Reviewer 4: I think we said `expose` would be better here.

Author: fixed.

> `kyma function get <function-name> -n <namespace>`

Author: nice idea

*Expose*

`kyma function expose <function-name> --secure(optional) --actions <GET/POST> -n <namespace>`

*Service Catalog*

> `kyma function bind <function-name> --binding-instaces [<binding-instance-name>,...] -n <namespace>`

Reviewer 2: Shouldn't it be a service-class instead of a binding-instance? 🤔

`kyma function unbind <function-name> --binding-usage [<binding-usage-name>,...] -n <namespace>`

*Bind to events*

`kyma function trigger <function-name> --events [<event-name>,..] -n <namespace>`
`kyma function trigger <function-name> --events [<event-name>,..] -n <namespace>`

*Bind to environment variables*

`kyma function bind <function-name> --env [<foo=bar>,..] -n <namespace>`
`kyma function unbind <function-name> --env [<foo=bar>,..] -n <namespace>`

*List all functions*

`kyma function list -n <namespace>`

> A detailed table view with function name, runtime, age, replicas and state

*Runtimes*

`kyma function get runtimes`

*Logs*

> `kyma function <function-name> logs -n <namespace>`

Reviewer 4: ```suggestion
`kyma function logs <function-name> -n <namespace>`
```

Author: fixed.

> `kyma function get runtimes`

Reviewer 2: In `kyma function init <function-name> --language <language-name>` language is used instead of runtime. Is this on purpose?

Author: fixed.

*Logs*

`kyma function logs <function-name> -n <namespace>`

*function Status*

`kyma function status <function-name> -n <namespace>`

Should display status

```bash
  Name: <function-name>
  Runtime: <my-runtime>
  State: Running/Error/Deploying/...
  Replicas: 1
  Min-replicas: 1
  Max-replicas: 2
  API: <url>
    Secure: yes/no
    Actions: GET, POST
  Bindings:
    Events: <list of events>
    Servicebindings: <list of service bindings>
  labels: foo:bar
```

`kyma function show <function-name> events`
`kyma function show <function-name> apis`
`kyma function show <function-name> service-binding-usages`
`kyma function show <function-name> labels`
`kyma function show <function-name> replicas`

#### Typical User flow

This section details a typical development flow for the functions

> 1. New User creating function from scratch or from existing github project as shown [here](#creation-of-project-locally). This would generate a function project for vscode. User is also provided with readme file to explain the structure of project and also the config.yaml. It should also explain how user can test it locally and also debugging.

Reviewer 2: Would work in general with git I guess:

```suggestion
1. New User creating function from scratch or from existing git project as shown [here](#creation-of-project-locally). This would generate a function project for vscode. User is also provided with readme file to explain the structure of project and also the config.yaml. It should also explain how user can test it locally and also debugging.
```

Author: removed this.

> 1. New User creating function from scratch or from existing github project as shown [here](#creation-of-project-locally). This would generate a function project for vscode. User is also provided with readme file to explain the structure of project and also the config.yaml. It should also explain how user can test it locally and also debugging.

Reviewer 2: What does it mean to use `--from-git` ? Will it just do a `git clone` and then bootstrap the function or do you have more in your mind ?

Author: removed this

3. User can change the code and test it locally by running a server locally on his workstation. User can make use of make file provided.

4. For debugging purpose user can either run it locally and attach the debuger or for in-cluster debugging he can use the `kyma function debug` command which would use telepresence to redirect the event triggers for example to the local code and he can attach the debugger to debug.

### Service catalog

We should enable kyma cli to be able to configure service catalog too. We should support following commands. The implementation of the commands should be abstracted in a separate library:

`kyma service-instance list`

#### Service Bindings

##### CRUD

`kyma service-binding create <service-binding-name> -n <namespace> --service-instance <service-instance-name>`

`kyma service-binding delete <service-binding-name> -n <namespace>`

`kyma service-binding get <service-binding-name> -n <namespace>`

#### Service Binding Usage

##### CRUD

`kyma service-binding-usage create <sbu-name> -n namespace --service-binding <service-binding-name> --function <function-name>`

> `kyma service-binding-usage create <sbu-name> -n namespace --service-binding <service-binding-name> --microservice <microservice-name>`

Reviewer 2: So far in the proposal kyma cli was not mentioning microservices. If we support them as well in addition to functions, shouldn't we have a kyma microservice command as well? Or can the function command be used for it somehow?

Author: service catalog should be discussed in detail.

`kyma service-binding-usage update <sbu-name> -n namespace --service-binding <service-binding-name> --function <function-name>`

`kyma service-binding-usage update <sbu-name> -n namespace --service-binding <service-binding-name> --microservice <microservice-name>`

`kyma service-binding-usage delete <sbu-name> -n namespace`

> `kyma service-binding-usage delete <sbu-name> -n namespace`
> 
> `kyma service-binding-usage get <sbu-name> -n namespace`
> 

Reviewer 2: Duplicate
```suggestion
```

`kyma service-binding-usage get <sbu-name> -n namespace`

### Events

We should enable support for events in kyma cli too. We should have support for following commands and implementation abstracted as a separate library:

`kyma event list -n <namespace>`

> `kyma event create -n <namespace> subscriptions --topic {"foo":"bar"}`

Reviewer 2: What does topic `{"foo":"bar"}` mean? Can you elaborate more?

Author: lets discuss this later in detail.

# Managing Kyma components using Kyma CLI

Kyma cli has been used for cluster management. We should look into how we can introduce support for management of components on the kyma cluster. This would bring in following benefits:

* Enable users who have limited or no kuberenetes knowledge to use kyma.
* Provide an abstraction on complex kubectl command chains.
* Could be easily used into CI/CD pipelines where they can have automated way of configuring kyma cluster to have functions deployed via CI/CD pipeline

## Architecture

We should follow the modular approach with each component having its own library. This library can be used inside the kyma cli for acessing various components. The library contains all the implementation. The kyma cli should invoke this library. We can take inspiration from hydroform and implement it in similar way. Currently we have following modules in kyma:

  1. Functions
  2. Service Catalog
  3. Events
  4. Application Connector
  5. APIs

## Command structure

The command structure used by cloud providers is following:

```bash
<cliName> <functionality> <action>
#example aws s3 ls

# Functions

> kyma lambda <action>

Reviewer: rename to function

Author: fixed.

##### Creation of project locally

This could be used for the user who is new to kyma and kubernetes in general. He should be able to have an example function created easily. We can have following commands.

```bash
  kyma function init <function-name> --language <language-name>

  kyma function init <function-name> --language <language-name> -n <namespace> -p <path>

  kyma function init <function-name> --language <language-name> --from-git <path-to-function-code>
```

> The namespace should be optional. When not passed it can be set to `default`. When the user has path to the function code (residing in git) we should generate the scaffolding around it (if its not existing already).
When path `-p` is passed then it should create the project in the designated path, otherwise, it uses the current directory.
The language corresponds the language in which the function code would be written.

This should create directoy structure with following content based on the language passed (assuming vscode as the IDE).:

```bash
  <function-name>
  ├── .vscode
  │   └── launch.json
  ├── deployment
  │   └── deployment.yaml
  ├── local
  │   └── <server-code.extension>
  └── Readme.md

>   └── code

Reviewer 5: just naming, but `src` is more commonly used as directory for source code rather than `code`

    ├── config.yaml
    ├── <function-code.extension>
    └── <dependecies.extension>
```

* launch.json consists of settings for the local debugger on vscode

> * deployment.yaml consists of the yaml of the function cr that has been deployed on the kyma cluster. This would be generated once function has been deployed.

Reviewer: Isn't deployment.yaml still misleading? Maybe function.yaml?

Reviewer: The file will not be used later, it is just for visibility or a dry-run?

Author: It is to show the user what has been deployed onto the cluster.

Developer can modify this file to deploy or update a function on to the kyma cluster. This would enable the developer not only tune the function (by modifying replica, memory cpu), but also provide an overview of the current triggers for the function.

Usually one does not change the config so often during development compared to code. So our aim should be to make deployment and testing the function simpler. So having a config file taking care of configuration would make deployment much simpler.

Having such a config file would help developers with limited or no kubernetes knowledge to get themselves aquainted with various configuration options for function and concentrate on the business logic for writing the functions.

Additionally it can be used automated deployments of functions. As this file can be read by kyma cli to deloy function with required configurations.

Example structure of config.yaml

```yaml
name: <function_name>
namespace: <namespace> #default by default
events:
  - event1
  - event2
environment-vars:
  - foo:bar

> api:

Reviewer: So it will also define APIRules. Sounds nice, but am afraid to support here the full feature set, think of all the different auth options. So either we limit it to a very specific minimal feature set or we have a generic mechanism in place which mainly takes the content here and aplies it 1:1 to the apiserver. What is the plan?

Author: APIRules would be implemented as a library. The feature set would be defined what has been implemented inside the library.

Reviewer 2: (1) If we implement something like this should we give the user the possibility to output the yamls which would be used to realize the function defined in `config.yaml`. Same as with `deployment/function.yaml` ?
(2) I like the idea, but that sounds like another API for functions (and also eventing, authentication etc.). What about we translate this config file or some kyma cli invocation (e.g. ```kyma function deploy --namespace <namespace> --name <name> --trigger-event stage.com.external.solution.order.created ...```) to the actual kubernetes CRs and put them inside the `deployment` folder ?  So not using a `config.yaml` at all.

`kyma function debug <function-name> -n <namespace>`

> We can use telepresence for local debugging. We can either use --swap-deployment or --new-deployment option.

Telepresence would create a proxy in the kyma cluster and redirect the calls to local codebase. One can attach the debugger to his code base along with breakpoints to debug the code.

##### Deploying and running the project

The function generated by `init` command should be deployable onto kyma cluster. For deploying the file onto kyma cluster it would read the values from `config.yaml`. If the api name is not set then we should show some warning message and deploy.

For running/debugging it locally we should provide makefile. The running/debugging commands should be part of readme.

#### Working with deployed kyma cluster

We should allow deployment of the functions on a k8s cluster. Following commands should be supported

##### CRUD

##### Commands that use config

> `kyma function create -p <path-to-function-dir>`

Reviewer 5: Why do you think it is important to distinguish between "create" and "update" here? Shouldn't a `kubectl apply` way of deploying the function to the cluster be sufficient? Otherwise the developer (or CI system) needs to build logic to check if it is existing and so on.

Reviewer 5: From a naming perspective i think terms like `apply` or `deploy` or possibly `install` describe more explicitly that this command will transfer the local state to the cluster rather than `create`/`update`.

Author: Yeah apply sounds better here
